mixin/smarty-v2@1/mixin.php

```diff
@@ -23,9 +23,15 @@
   }
 
   $register = function() use ($dir) {
-    // This implementation has a theoretical edge-case bug on older versions of CiviCRM where a template could
-    // be registered more than once.
-    CRM_Core_Smarty::singleton()->addTemplateDir($dir);
+    // getTemplateDir returns string or array by reference
+    $templateRef = CRM_Core_Smarty::singleton()->getTemplateDir();
+    // Dereference and normalize as array
+    $templateDirs = (array) $templateRef;
+    // Add the dir if not already present
+    if (!in_array($dir, $templateDirs, TRUE)) {
+      array_unshift($templateDirs, $dir);
+      CRM_Core_Smarty::singleton()->setTemplateDir($templateDirs);
```

Comment:
How badly does it need the patch to Smarty 2 for this? That part wouldn't be portable. (And that's the part that probably needs portability most?)

In smarty2, the `$template_dirs` var is public, so, you could so something like:

```php
if (smells like smarty 2) {
  CRM_Core_Smarty::singleton()->template_dirs = $templateDirs;
}
else {
  CRM_Core_Smarty::singleton()->setTemplateDir($templateDirs);
}
```

Reply:
@totten well, I guess we could, but that begs the question of why wasn't that a problem before? Because the "before" version of this patch already used the function `addTemplateDir()` which also doesn't natively exist in Smarty2 and was only recently added.

Reply:
I don't completely follow the problem. FWIW, `CRM_Core_Smarty::addTemplateDir()` has been around since at least [4.3](https://github.com/civicrm/civicrm-core/blob/4.3/CRM/Core/Smarty.php#L218).

Reply:
Ok yea I see that now, it was recently moved but it did exist before.